decidim-participatory_processes/lib/decidim/participatory_processes/query_extensions.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+module Decidim
+  module ParticipatoryProcesses
+    # This module's job is to extend the API with custom fields related to
+    # decidim-participatory_processes.
+    module QueryExtensions
```

Comment:
I don't think you need `QueryExtensions` for this. It's quite obscure, but we already [define fields for all participatory spaces](https://github.com/decidim/decidim/blob/master/decidim-core/lib/decidim/query_extensions.rb#L16).

Reply:
I know, it took me a while to understand how you've done it. However, I also seen that not all the participatory spaces share the exact same fields. Specially the `slug` field (ie initiatives doesn't have it, at least as a entry in the database). That's why I decided to overwrite the participatory Process extensions (cause I think it's important to be able to search by slug). Do you think there's a better way to do it? thanks

Reply:
Oh, I suspected it might be that! We could proceed in improving this in two ways:

* Allowing each participatory space to optionally override its field on the API in order to accommodate different find behaviors, and resorting to a default way otherwise.
* (probably better) make slug be mandatory to all participatory spaces. Slugs are a good UX practice and are also great for SEO purposes.

What do you think?

Reply:
I suspect that the second approach is going to take more time and discussions, because it will affect some parts of the code already working.
Overwrite is always available, and we can use it meanwhile the second approach is not implemented or in evaluation. However if you think there's a better way to do that overwriting, please let me know and I'll do it.

Reply:
I agree with you. I don't think to override the way you did is the best idea though, since it's kind of obscure (the same thing is done twice along the codebase). Specifying the field [somewhere in the participatory space contract](https://github.com/decidim/decidim/blob/master/decidim-participatory_processes/lib/decidim/participatory_processes/participatory_space.rb#L11) would possibly be better, as it expresses the intent more clearly.

So, IMHO, we should remove these `QueryExtensions` right now, and figure out a way to do it like this in another PR.

Reply:
You mean, like adding a new item in [that space contract file](https://github.com/decidim/decidim/blob/master/decidim-participatory_processes/lib/decidim/participatory_processes/participatory_space.rb#L11), for example:
```ruby
  participatory_space.query_type = "Decidim::ParticipatoryProcesses::ParticipatoryProcessType"
  participatory_space.query_by = [:id, :slug]
```
And, also the corresponding property in in the [manifest class file](https://github.com/decidim/decidim/blob/master/decidim-core/lib/decidim/participatory_space_manifest.rb#L25).

Then use this (if present) in the [core query_extensions file](https://github.com/decidim/decidim/blob/master/decidim-core/lib/decidim/query_extensions.rb#L36) to add more searchable fields if the `query_by` array is defined (if empty would default to `id`). This way we could provide a simple method to specify how to get a space by simply listing them in the manifest.

Instead of an array, we could use a hash to allow different fields in the graphql and the database (although I don't see an immediate use for that).

What do you think?

Reply:
Sorry for being unresponsive these last few days!

Your approach is quite great, although it might not be flexible enough for your needs. What about just allowing overriding the whole finder? Something like:

```ruby
participatory_space.extend_query_field do
  argument :slug, types.String, "The Slug of the ParticipatoryProcess"
  resolve lambda { |_obj, args, ctx|
    q = { organization: ctx[:current_organization] }
    args[:slug] ? q[:slug] = args[:slug] : q[:id] = args[:id]
    ParticipatoryProcess.public_spaces.find_by(q)
  }
end
```

Reply:
Looks great to me!

Sorry if I am asking something silly, just want to be sure to make things right:

Would that `extend_query_field` be optional right? so by default would follow the original implementation and I still would have to modify the *space contract file*, *the manifest and the core query extension files* as I pointed in my previous comment.

If that's ok, I'll update the pull request and we can continue the review then.

Reply:
Exactly, that's what I was thinking: Having a default but allowing optionally overriding it. Sorry if I'm being too nitpicky here, but it's always better to find contract and abstractions whenever possible :)

Thanks for the work!

Reply:
Maybe we could do this in a further PR and just merge the current PR with the specs and fields (and leaving out the QueryExtensions)?

Reply:
yes, you're right, it will be easier to manage smaller PRs. Hopefully I'll have time to do that today.
Thanks for your patience!